### ISSUE TITLE
Lost Type Variable after applying generic function

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3208]
**Describe the bug**
The code below has two unknowns where I would expect the type variable to be preserved. I'm less confident on this report due to ambiguity of type variable scoping. I did check that mypy does preserve the type variables for the two examples.

```python
from typing import Callable, Generic, List, Type, TypeVar

T1 = TypeVar("T1")
T2 = TypeVar("T2")


class Foo(Generic[T1]):
    ...


def map_foo(foo: Foo[T1], f: Callable[[T1], T2]) -> Foo[T2]:
    ...


def map_foo_type(foo: Type[Foo[T1]], f: Callable[[T1], T2]) -> Type[Foo[T2]]:
    ...


result1 = map_foo(Foo[List[T1]](), set)
reveal_type(result1)  # Foo[set[Unknown]]. The TypeVar is lost.

result2 = map_foo_type(Foo[List[int]], lambda ls: set(ls))
reveal_type(result2)  # Type[Foo[set[unknown]]]. The TypeVar is lost.
```

**Expected behavior**
result1 should be Foo[set[T1]] while result2 should be Type[Foo[set[T1]]].

**VS Code extension or command-line**
pyright cli 1.1.230

**Additional context**
I would also be open to alternative ways to write code like this if this is a type var ambiguity issue.

edit: I just realized maybe this is higher kinded type related. Not sure if that's required for this example or not. I'm confused on why mypy preserves the types here and if it just happens to have some certain cases of HKT supported. If the issue is about HKT then probably should be closed.

**Comment by erictraut** [comment]
I'm not able to repro the behavior you're reporting in either pyright or mypy.

For me, pyright correctly preserves the TypeVar in the second case (result2) but does not in the first case (result1). I see the same behavior in mypy.

In your first case, you're using `T1` in a context where it has no meaning. There is no class, function or type alias scope that it associated with. So the behavior is expected.

**Comment by hmc-cs-mdrissi** [comment]
Sorry on mistake with bug report. I should have tested more carefully.

I agree on first case. For second case I realized why I saw something different. I originally had this variant,

```python
from typing import Callable, Generic, List, TypeVar

T1 = TypeVar("T1")
T2 = TypeVar("T2")

class Foo(Generic[T1]):
    ...

def map_foo(foo: Foo[T1], f: Callable[[T1], T2]) -> Foo[T2]:
    ...

result1 = map_foo(Foo[List[int]](), set)
reveal_type(result1)  # Foo[set[Unknown]]. The TypeVar is lost.
```

But a small variant, `result1 = map_foo(Foo[List[int]](), lambda ls: set(ls))` works. Mypy and pyright are consistent here and both having missing type argument. So maybe this is not a type checker issue. I'm confused why `set` vs `lambda ls: set(ls) `differ although my best guess is this is a typeshed detail I'm missing. Or that `f` vs `lambda x: f(x)` aren't always same.

**Comment by erictraut** [comment]
The problem with the second example is that `map` and `set` are both generic classes. When you call the `map` constructor, the type evaluator is "solving" for the type variables that are in scope for the `map` class. It is not solving the type variables in scope for the `set` class. You would need to specify the type arguments for `set` explicitly in this case: `y = map(set[int], x)`.

The first example works because there are two calls. The first is a call to the `map` constructor, and the second is a call to the `set` constructor.

In short, a call expression solves the type variables for the class or function that is being called, not type variables associated with other scopes.
